### PR TITLE
Deployment script prompt

### LINKED ISSUE #224: Improve performance of deploy
I'll have to think through the experience for making this work, but I want to make sure the need is captured. This will ensure Windows users are able to deploy compiled modules.

Things to consider:

* An app could be one of many runtimes (and may be custom)
* How can we know to alert a user that they need to enable npm install on deploy?
* At what point in the process should this be configured?

See also [vscode-azuretools#7](https://github.com/Microsoft/vscode-azuretools/issues/7) and the [Kudu docs](https://github.com/projectkudu/kudu/wiki/Configurable-settings#enabledisable-build-actions)

### PR DESCRIPTION
**Pull request by nturinski**
Fixes #224

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
     "name": "vscode-azureappservice",
     "displayName": "Azure App Service",
     "description": "An Azure App Service management extension for Visual Studio Code.",
-    "version": "0.5.1",
+    "version": "0.6.0",
```
Comment (EricJizbaMSFT) [inline package.json]: nit: I prefer you bump the version in the same PR that you do the changelog

File: package.json [hunk 2]
```diff
@@ -490,18 +490,19 @@
...
-                    "default": "",
+                    "default": [],
                     "description": "Defines which files in the workspace to ignore for Zip deploy. This applies to Zip deploy only, has no effect on other deployment methods."
                 },
                 "appService.showRemoteFiles": {
                     "type": "boolean",
                     "default": false,
                     "description": "Show or hide the File Nodes of a Web App.  Caution: Manually editing your Web App's files could cause unexpected behavior."
                 },
-                "appService.showSavePrompt": {
```
Comment (EricJizbaMSFT) [inline package.json]: Don't delete the 'showSavePrompt' setting

File: package.json [hunk 3]
```diff
@@ -490,18 +490,19 @@
...
+                "appService.neverPromptBuildDuringDeploy": {
+                    "scope": "resource",
```
Comment (EricJizbaMSFT) [inline package.json]: This flag doesn't match the other flags. Do something like "showBuildDuringDeployPrompt"

File: src/explorer/DeploymentSlotTreeItem.ts [hunk 4]
```diff
@@ -31,7 +32,7 @@ export class DeploymentSlotTreeItem extends SiteTreeItem {
     }
 
     public async loadMoreChildren(_node: IAzureParentNode<DeploymentSlotTreeItem>): Promise<IAzureTreeItem[]> {
-        return workspace.getConfiguration().get("appService.showRemoteFiles") ?
+        return workspace.getConfiguration('appService').get(configurationSettings.showRemoteFiles) ?
```
Comment (EricJizbaMSFT) [inline src/explorer/DeploymentSlotTreeItem.ts]: nit: you could do this in contants.ts:
```
export const extensionPrefix: string = 'appService';
```

File: src/explorer/SiteTreeItem.ts [hunk 5]
```diff
@@ -124,6 +136,33 @@ export abstract class SiteTreeItem implements IAzureParentTreeItem {
             });
     }
 
+    private async enableScmDoBuildDuringDeploy(fsPath: string, runtime: string): Promise<void> {
+        const yesButton: MessageItem = { title: 'Yes' };
+        const noButton: MessageItem = { title: 'No', isCloseAffordance: true };
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: Use 'Dont' warn again' to be consistent with our other dialogs. It also makes it more clear that this action is different than just closing the dialog

File: src/explorer/SiteTreeItem.ts [hunk 6]
```diff
@@ -124,6 +136,33 @@ export abstract class SiteTreeItem implements IAzureParentTreeItem {
...
+                opn('https://aka.ms/Kwwkbd');
+            }
+        }
+        if (input === yesButton) {
+            let oldSettings: string[] | string = workspace.getConfiguration('appService', Uri.file(fsPath)).get(constants.configurationSettings.zipIgnorePattern);
+            if (typeof oldSettings === "string") {
+                oldSettings = [oldSettings];
+                // settings have to be an array to concat the proper zipIgnoreFolders
+            }
+            workspace.getConfiguration('appService', Uri.file(fsPath)).update(constants.configurationSettings.zipIgnorePattern, oldSettings.concat(zipIgnoreFolders), ConfigurationTarget.WorkspaceFolder);
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: Don't pass in `ConfigurationTarget.WorkspaceFolder`. That's already the default

File: src/explorer/SiteTreeItem.ts [hunk 7]
```diff
@@ -124,6 +136,33 @@ export abstract class SiteTreeItem implements IAzureParentTreeItem {
...
+            await fse.writeFile(path.join(fsPath, constants.deploymentFileName), constants.deploymentFile);
+        } else {
+            workspace.getConfiguration('appService', Uri.file(fsPath)).update(constants.configurationSettings.neverPromptBuildDuringDeploy, true, ConfigurationTarget.WorkspaceFolder);
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: Only update this if they click "Don't warn again". If they close the dialog, we should warn again

File: src/explorer/SiteTreeItem.ts (hunk 5 above)
Comment (nturinski) [reply src/explorer/SiteTreeItem.ts]: I did 'Don't show again' since it's not really a warning

File: src/explorer/SiteTreeItem.ts [hunk 8]
```diff
@@ -124,6 +136,33 @@ export abstract class SiteTreeItem implements IAzureParentTreeItem {
             });
     }
 
+    private async enableScmDoBuildDuringDeploy(fsPath: string, runtime: string): Promise<void> {
+        const yesButton: MessageItem = { title: 'Yes' };
+        const dontShowAgainButton: MessageItem = { title: "Don't show again", isCloseAffordance: true };
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: This button isn't the close affordance. There should still be a 'close' button that means "I don't want to enable scm right now, but warn me again next time"

File: src/explorer/SiteTreeItem.ts [hunk 9]
```diff
@@ -124,6 +136,33 @@ export abstract class SiteTreeItem implements IAzureParentTreeItem {
...
+        const learnMoreButton: MessageItem = { title: 'Learn More' };
+        const zipIgnoreFolders: string[] = constants.getIgnoredFoldersForDeployment(runtime);
+        const buildDuringDeploy: string = `Run build script during deployment?  The "${zipIgnoreFolders.join(", ")}" directory will be built during the deployment, rather than zipped, resulting in a faster deployment.`;
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: How about "Would you like to configure your project for faster deployment?"

I'm worried that we'll just confuse users if we include too much of the "how" in the message, and I think that's what the "learn more" button is for.

**Pull request merged**